File: leetcode.com/0905_sort_array_by_parity/solution.go
package main

import "fmt"

func sortArrayByParity(A []int) []int {
    var sorted = make([]int, len(A))
    var begin int
    var end int

    for _, value := range A {
        if value % 2 == 0 && end <= begin {
            sorted[end] = value
            end++;
            begin++
        } else if value % 2 != 0 {
            sorted[end] = value
            end++
        } else {
            tmp := sorted[begin]
            sorted[begin] = value
            sorted[end] = tmp
            begin++
            end++
        }
    }

    return sorted
}

func main() {
    fmt.Println(sortArrayByParity([]int{0,1,2}))
}

File: leetcode.com/0905_sort_array_by_parity/solution.c
int* sortArrayByParity(int* A, int ASize, int* returnSize)
{
    int* sorted = malloc(sizeof(int) * ASize);
    int begin = 0;
    int end = 0;
    
    for (auto int i = 0 ; i < ASize ; i++) {
        if (A[i] % 2 == 0 && end <= begin) {
            sorted[end++] = A[i];
            begin++;
        } else if (A[i] % 2 != 0) {
            sorted[end++] = A[i];
        } else {
            int tmp = sorted[begin];
            sorted[begin++] = A[i];
            sorted[end++] = tmp;
        }
    }
    
    *returnSize = ASize;
    return sorted;
}


